airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/VersionedAirbyteStreamFactory.java

```diff
@@ -43,6 +61,91 @@ public VersionedAirbyteStreamFactory(final AirbyteMessageSerDeProvider serDeProv
+    Preconditions.checkNotNull(protocolVersion);
+    this.serDeProvider = serDeProvider;
+    this.migratorFactory = migratorFactory;
+    this.initializeForProtocolVersion(protocolVersion);
+  }
+
+  /**
+   * Create the AirbyteMessage stream.
+   *
+   * If detectVersion is set to true, it will decide which protocol version to use from the content of
+   * the stream rather than the one passed from the constructor.
+   */
+  @SneakyThrows
+  @Override
+  public Stream<AirbyteMessage> create(final BufferedReader bufferedReader) {
+    if (shouldDetectVersion) {
+      final Optional<Version> versionMaybe = detectVersion(bufferedReader);
+      if (versionMaybe.isPresent()) {
+        logger.info("Detected Protocol Version {}", versionMaybe.get().serialize());
+        initializeForProtocolVersion(versionMaybe.get());
+      } else {
+        // No version found, use the default as a fallback
+        logger.info("Unable to detect Protocol Version, assuming protocol version {}", fallbackVersion.serialize());
+        initializeForProtocolVersion(fallbackVersion);
+      }
+    }
+    return super.create(bufferedReader);
+  }
+
+  /**
+   * Attempt to detect the version by scanning the stream
+   *
+   * Using the BufferedReader reset/mark feature to get a look-ahead. We will attempt to find the
+   * first SPEC message and decide on a protocol version from this message.
+   *
+   * @param bufferedReader the stream to read
+   * @return The Version if found
+   * @throws IOException
+   */
+  private Optional<Version> detectVersion(final BufferedReader bufferedReader) throws IOException {
+    // Buffersize needs to be big enough to containing everything we need for the detection. Otherwise,
+    // the reset will fail.
+    bufferedReader.mark(BUFFER_READ_AHEAD_LIMIT);
+    try {
+      // Cap detection to the first 10 messages. When doing the protocol detection, we expect the SPEC
+      // message to show up early in the stream. Ideally it should be first message however we do not
+      // enforce this constraint currently so connectors may send LOG messages before.
+      for (int i = 0; i < MESSAGES_LOOK_AHEAD_FOR_DETECTION; ++i) {
+        final String line = bufferedReader.readLine();
+        final Optional<JsonNode> jsonOpt = Jsons.tryDeserialize(line);
+        if (jsonOpt.isPresent()) {
+          final JsonNode json = jsonOpt.get();
+          if (isSpecMessage(json)) {
+            final JsonNode protocolVersionNode = json.at("/spec/protocol_version");
+            bufferedReader.reset();
+            return Optional.ofNullable(protocolVersionNode).filter(Predicate.not(JsonNode::isMissingNode)).map(node -> new Version(node.asText()));
+          }
+        }
+      }
+      bufferedReader.reset();
+      return Optional.empty();
+    } catch (IOException e) {
+      logger.warn(
+          "Protocol version detection failed, it is likely than the connector sent more than {}B without an complete SPEC message." +
```

Comment:
Is the `B` after `{}` a typo or is it meant to represents bytes?

Reply:
It is meant for bytes.